crates/net/src/discv4.rs

```diff
@@ -0,0 +1,330 @@
+            ip,
+            udp_port,
+            tcp_port,
+        };
+        Ok((endpoint, remaining))
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(crate) struct PingMessage {
+    /// The Ping message version. Should be set to 4, but mustn't be enforced.
+    version: u8,
+    /// The endpoint of the sender.
+    from: Endpoint,
+    /// The endpoint of the receiver.
+    to: Endpoint,
+    /// The expiration time of the message. If the message is older than this time,
+    /// it shouldn't be responded to.
+    expiration: u64,
+    /// The ENR sequence number of the sender. This field is optional.
+    enr_seq: Option<u64>,
+}
+
+impl PingMessage {
+    pub fn new(from: Endpoint, to: Endpoint, expiration: u64) -> Self {
+        Self {
+            version: 4,
+            from,
+            to,
+            expiration,
+            enr_seq: None,
+        }
+    }
+
+    // TODO: remove when used
+    #[allow(unused)]
+    pub fn with_enr_seq(self, enr_seq: u64) -> Self {
+        Self {
+            enr_seq: Some(enr_seq),
+            ..self
+        }
+    }
+}
+
+impl RLPEncode for PingMessage {
+    fn encode(&self, buf: &mut dyn BufMut) {
+        match self.enr_seq {
+            Some(seq) => (self.version, &self.from, &self.to, self.expiration, seq).encode(buf),
+            None => (self.version, &self.from, &self.to, self.expiration).encode(buf),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(crate) struct PongMessage {
+    /// The endpoint of the receiver.
+    to: Endpoint,
+    /// The hash of the corresponding ping packet.
+    ping_hash: H256,
+    /// The expiration time of the message. If the message is older than this time,
+    /// it shouldn't be responded to.
+    expiration: u64,
+    /// The ENR sequence number of the sender. This field is optional.
+    enr_seq: Option<u64>,
+}
+
+impl PongMessage {
+    #[allow(unused)]
+    pub fn new(to: Endpoint, ping_hash: H256, expiration: u64) -> Self {
+        Self {
+            to,
+            ping_hash,
+            expiration,
+            enr_seq: None,
+        }
+    }
+
+    // TODO: remove when used
+    #[allow(unused)]
+    pub fn with_enr_seq(self, enr_seq: u64) -> Self {
+        Self {
+            enr_seq: Some(enr_seq),
+            ..self
+        }
+    }
+}
+
+impl RLPDecode for PongMessage {
+    fn decode_unfinished(rlp: &[u8]) -> Result<(Self, &[u8]), RLPDecodeError> {
+        let decoder = Decoder::new(rlp)?;
+        let (to, decoder) = decoder.decode_field("to")?;
+        let (ping_hash, decoder) = decoder.decode_field("ping_hash")?;
+        let (expiration, decoder) = decoder.decode_field("expiration")?;
+        let (enr_seq, decoder) = decoder.decode_optional_field();
+
+        let pong = PongMessage::new(to, ping_hash, expiration);
+
+        let pong = if let Some(seq) = enr_seq {
+            pong.with_enr_seq(seq)
+        } else {
+            pong
+        };
```

Comment:
We should instantiate the struct directly to avoid matching the `Option`.

Reply:
Done!